librz/core/cmd/cmd_debug.c

```diff
@@ -4410,24 +4406,58 @@ RZ_IPI RzCmdStatus rz_cmd_debug_add_bp_handler(RzCore *core, int argc, const cha
 
 // dbl
 RZ_IPI RzCmdStatus rz_cmd_debug_list_bp_handler(RzCore *core, int argc, const char **argv, RzCmdStateOutput *state) {
-	RzOutputMode mode = state->mode;
-	switch (mode) {
-	case RZ_OUTPUT_MODE_STANDARD:
-		rz_bp_list(core->dbg->bp, 0);
-		break;
-	case RZ_OUTPUT_MODE_RIZIN:
-		rz_bp_list(core->dbg->bp, 1);
-		break;
-	case RZ_OUTPUT_MODE_JSON:
-		rz_bp_list(core->dbg->bp, 'j');
-		break;
-	case RZ_OUTPUT_MODE_QUIET:
-		rz_bp_list(core->dbg->bp, -1);
-		break;
-	default:
-		rz_warn_if_reached();
-		break;
+	rz_return_val_if_fail(state, RZ_CMD_STATUS_ERROR);
+	RzBreakpointItem *b;
+	RzListIter *iter;
+	PJ *pj = NULL;
+	pj = state->d.pj;
+	rz_cmd_state_output_array_start(state);
+	rz_list_foreach (core->dbg->bp->bps, iter, b) {
+		switch (state->mode) {
+		case RZ_OUTPUT_MODE_STANDARD:
+			rz_cons_printf("0x%08" PFMT64x " - 0x%08" PFMT64x
+				       " %d %s %s %s %s %s cmd=\"%s\" cond=\"%s\" "
+				       "name=\"%s\" module=\"%s\"\n",
+				b->addr, b->addr + b->size, b->size,
+				rz_str_rwx_i(b->perm),
+				b->hw ? "hw" : "sw",
+				b->trace ? "trace" : "break",
+				b->enabled ? "enabled" : "disabled",
+				rz_bp_is_valid(core->dbg->bp, b) ? "valid" : "invalid",
+				rz_str_get(b->data),
+				rz_str_get(b->cond),
+				rz_str_get(b->name),
+				rz_str_get(b->module_name));
+			break;
+		case RZ_OUTPUT_MODE_RIZIN:
+			if (b->module_name) {
+				rz_cons_printf("dbm %s %" PFMT64d "\n", b->module_name, b->module_delta);
+			} else {
+				rz_cons_printf("db @ 0x%08" PFMT64x "\n", b->addr);
+			}
+			break;
+		case RZ_OUTPUT_MODE_JSON:
+			pj_o(pj);
+			pj_kN(pj, "addr", b->addr);
+			pj_ki(pj, "size", b->size);
+			pj_ks(pj, "perm", rz_str_rwx_i(b->perm & 7)); // filter out only rwx bits
```

Comment:
```suggestion
			pj_ks(pj, "perm", rz_str_rwx_i(b->perm));
```

Reply:
I don't really see how the higher bits (`RZ_PERM_(SHAR|PRIV|ACCESS|CREAT)`) make sense in the context of breakpoints. Shouldn't we just discard them?

Reply:
i do not think you ever set them, so they cannot be set. if that is an issue, just `b->perm & 7` every time you call `rz_str_rwx_i`

Reply:
Yeah, I filter once while setting the perms. No need to filter while printing again. Agreed.